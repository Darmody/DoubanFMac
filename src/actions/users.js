// @flow
import * as types from 'constants/types/ActionTypes'
import type { Action } from 'constants/types/Redux'

export const current = (): Action => ({
  type: types.USER_CURRENT_REQUEST,
  payload: undefined,
})
