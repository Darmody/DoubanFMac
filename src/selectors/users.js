// @flow
import { USER } from 'schemas'
import type { StateShape } from 'constants/types/Redux'
import { selectEntity } from 'selectors/entities'

export const selectCurrent = (state: StateShape) => selectEntity(
  state, USER, state.auth.id,
)
