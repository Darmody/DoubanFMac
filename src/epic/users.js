// @flow
import { Observable as Rx$ } from 'rxjs'
import * as types from 'constants/types/ActionTypes'
import type { Epic } from 'constants/types/Redux'
import * as API from 'clients/doubanRxClient'
import { logout } from 'actions/auth'
import { listen } from 'actions/songs'
import { save as saveEntities } from 'actions/entities'
import { normalizeResponse, getToken, rejected } from 'utils/operators'
import { USER } from 'schemas'

const startListen = store => Rx$.of(listen(store.getState().channels.id))

const current: Epic = (action$, store) => action$
  .ofType(types.USER_CURRENT_REQUEST)
  .mergeMap(() => Rx$.concat(
    API
      .me(getToken(store))()
      .pluck('response')
      .map(normalizeResponse(USER))
      .map(({ entities }) => saveEntities(entities))
      .catch(error => Rx$.merge(
        rejected(types.USER_CURRENT_FAILURE)(error),
        Rx$.of(logout()),
      )),
    startListen(store),
  )
)

export default [
  current
]
