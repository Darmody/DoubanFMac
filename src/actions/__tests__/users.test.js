import * as actions from '../users'

describe('action::users', () => {
  it('创建一个 current action', () => {
    expect(actions.current()).toMatchSnapshot()
  })
})
