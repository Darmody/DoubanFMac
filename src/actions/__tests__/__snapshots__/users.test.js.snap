// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action::users 创建一个 current action 1`] = `
Object {
  "payload": undefined,
  "type": "USER/CURRENT/REQUEST",
}
`;
